File: source/aoImportWizard/Controllers/genericController.vb
﻿
Imports Contensive.ImportWizard.Models
Imports Contensive.BaseClasses

Namespace Contensive.ImportWizard.Controllers
    Public NotInheritable Class GenericController
        Private Sub New()
        End Sub
        '
        Public Shared Function normalizeFilename(srcFilename As String) As String
            Dim ext As String = IO.Path.GetExtension(srcFilename)
            Dim filenameNoExt As String = IO.Path.GetFileNameWithoutExtension(srcFilename)
            Dim result As String = ""
            Dim validCharacters As String = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
            For Each c In filenameNoExt.Trim().Replace(" "c, "_")
                result &= If(validCharacters.Contains(c), c, "-")
            Next
            Return If(String.IsNullOrEmpty(ext), result, result & "." & ext)
        End Function
        '
        Public Shared Function Base64Encode(ByVal plainText As String) As String
            Dim plainTextBytes = System.Text.Encoding.UTF8.GetBytes(plainText)
            Return System.Convert.ToBase64String(plainTextBytes)
        End Function
        '
        Public Shared Function Base64Decode(ByVal base64EncodedData As String) As String
            Dim base64EncodedBytes = System.Convert.FromBase64String(base64EncodedData)
            Return System.Text.Encoding.UTF8.GetString(base64EncodedBytes)
        End Function        '

        '
        '====================================================================================================
        ''' <summary>
        ''' if date is invalid, set to minValue
        ''' </summary>
        ''' <param name="srcDate"></param>
        ''' <returns></returns>
        Public Shared Function encodeMinDate(srcDate As DateTime) As DateTime
            Dim returnDate As DateTime = srcDate
            If srcDate < New DateTime(1900, 1, 1) Then
                returnDate = DateTime.MinValue
            End If
            Return returnDate
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' if valid date, return the short date, else return blank string 
        ''' </summary>
        ''' <param name="srcDate"></param>
        ''' <returns></returns>
        Public Shared Function getShortDateString(srcDate As DateTime) As String
            Dim returnString As String = ""
            Dim workingDate As DateTime = encodeMinDate(srcDate)
            If Not isDateEmpty(srcDate) Then
                returnString = workingDate.ToShortDateString()
            End If
            Return returnString
        End Function
        '
        '====================================================================================================
        Public Shared Function isDateEmpty(srcDate As DateTime) As Boolean
            Return (srcDate < New DateTime(1900, 1, 1))
        End Function
        '
        '====================================================================================================
        Public Shared Function getSortOrderFromInteger(id As Integer) As String
            Return id.ToString().PadLeft(7, "0"c)
        End Function
        '
        '====================================================================================================
        Public Shared Function getDateForHtmlInput(source As DateTime) As String
            If isDateEmpty(source) Then
                Return ""
            Else
                Return source.Year.ToString() + "-" + source.Month.ToString().PadLeft(2, "0"c) + "-" + source.Day.ToString().PadLeft(2, "0"c)
            End If
        End Function
        '
        '====================================================================================================
        Public Shared Function convertToDosPath(sourcePath As String) As String
            Return sourcePath.Replace("/", "\")
        End Function
        '
        '====================================================================================================
        Public Shared Function convertToUnixPath(sourcePath As String) As String
            Return sourcePath.Replace("\", "/")
        End Function
        '
        '   returns true if after removing this field, it is end of line
        '   Returns a cell from a csv source and advances the ptr to the start of the next field
        '   on entry, ptr points to the first character of the cell
        '   if there are spaces, they are included in the cell, unless the first non-space is a quote.
        '   if at end of line, the parseFieldReturnEol is true
        '   if end of file, return_eof is true
        '
        Public Shared Function parseFieldReturnEol(Source As String, sourcePtr As Integer, ByRef return_cell As String, ByRef return_ptr As Integer, ByRef return_eof As Boolean) As Boolean
            Try
                Dim result As Boolean
                '
                Dim crPtr As Integer
                Dim endPtr As Integer
                Dim Ptr As Integer
                Dim workingPtr As Integer
                Dim IsQuoted As Boolean
                Dim commaptr As Integer
                Dim lfPtr As Integer
                Dim crlfPtr As Integer
                Dim hint As String
                '
                Ptr = sourcePtr
                IsQuoted = False
                result = False
                '
                ' find initial character
                '
                hint = "no used"
                workingPtr = Ptr
                Do While Mid(Source, workingPtr, 1) = " "
                    workingPtr += 1
                Loop
                hint &= ",110"
                If Mid(Source, workingPtr, 1) = """" Then
                    '
                    ' if first non-space is a quote, ignore the leading spaces
                    '
                    Ptr = workingPtr
                    IsQuoted = True
                End If
                hint &= ",120"
                If Not IsQuoted Then
                    '
                    ' non-Quoted field
                    '
                    hint &= ",120"
                    commaptr = InStr(Ptr, Source, ",")
                    lfPtr = InStr(Ptr, Source, vbLf)
                    crPtr = InStr(Ptr, Source, vbCr)
                    crlfPtr = InStr(Ptr, Source, vbCrLf)
                    '
                    ' set workingPtr to the first one found
                    '
                    workingPtr = firstNonZero(commaptr, crlfPtr)
                    workingPtr = firstNonZero(workingPtr, lfPtr)
                    workingPtr = firstNonZero(workingPtr, crPtr)
                    'workingPtr = firstNonZero(workingPtr, commaptr)
                    workingPtr = firstNonZero(workingPtr, Len(Source))
                    If (workingPtr = crlfPtr) Then
                        '
                        ' end of line for crlf line
                        '
                        hint &= ",140"
                        endPtr = workingPtr - 1
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1)
                        return_ptr = workingPtr + 2
                        result = True
                    ElseIf (workingPtr = lfPtr) Then
                        '
                        ' end of line for lf line
                        '
                        hint &= ",130"
                        endPtr = workingPtr - 1
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1)
                        return_ptr = workingPtr + 1
                        result = True
                    ElseIf (workingPtr = crPtr) Then
                        '
                        ' end of line for cr line
                        '
                        hint &= ",135"
                        endPtr = workingPtr - 1
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1)
                        return_ptr = workingPtr + 1
                        result = True
                    ElseIf (workingPtr = commaptr) Then
                        '
                        ' end of cell, skip comma
                        '
                        hint &= ",150"
                        endPtr = workingPtr - 1
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1)
                        return_ptr = workingPtr + 1
                        result = False
                    Else
                        '
                        ' non of the above (non found) might be end of file
                        '
                        hint &= ",160"
                        endPtr = Len(Source)
                        If (endPtr - sourcePtr + 1 > 0) Then
                            return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1)
                        Else
                            return_cell = ""
                        End If
                        return_ptr = endPtr
                        result = True
                        result = True
                    End If
                Else
                    '
                    ' Quoted field, pass the initial quote
                    '
                    hint &= ",170"
                    Ptr += 1
                    Dim startPtr As Integer
                    startPtr = Ptr
                    '
                    Do While (Ptr <> 0) And (InStr(Ptr, Source, """") = InStr(Ptr, Source, """"""))
                        ' pass the doublequote
                        hint &= ",200"
                        Ptr = InStr(Ptr, Source, """""")
                        If (Ptr = 0) Then
                            '
                            ' neither quote or double quote were found - end of file error
                            '
                            hint &= ",210"
                            endPtr = Len(Source)
                            return_cell = ""
                            return_ptr = endPtr
                            result = True
                            result = True
                        Else
                            hint &= ",220"
                            Ptr += 2
                        End If
                    Loop
                    If Ptr <> 0 Then
                        '
                        ' ptr is on the closing quote
                        '
                        hint &= ",230"
                        Ptr = InStr(Ptr, Source, """")
                        endPtr = Ptr - 1
                        ' skip white space to next delimter
                        Do While (Mid(Source, Ptr + 1, 1) = " ") And (Ptr < Len(Source))
                            Ptr += 1
                        Loop
                        If (Ptr >= Len(Source)) Then
                            '
                            ' crlf end of line
                            '
                            hint &= ",240"
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1)
                            return_ptr = Ptr + 3
                            result = True
                        ElseIf (Mid(Source, Ptr + 1, 2) = vbCrLf) Then
                            ' ***** 20140131 - ptr is to the end quote
                            'If (Mid(Source, Ptr - 1, 2) = vbCrLf) Then
                            '
                            ' crlf end of line
                            '
                            hint &= ",250"
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1)
                            return_ptr = Ptr + 3
                            result = True
                        ElseIf (Mid(Source, Ptr + 1, 1) = vbLf) Then
                            '
                            ' lf end of line
                            '
                            hint &= ",240"
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1)
                            return_ptr = Ptr + 2
                            result = True
                        ElseIf (Mid(Source, Ptr + 1, 1) = vbCr) Then
                            '
                            ' cr end of line
                            '
                            hint &= ",240"
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1)
                            return_ptr = Ptr + 2
                            result = True
                        Else
                            '
                            ' not end of line, skip over anything before the next comma
                            '
                            hint &= ",260"
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1)
                            return_ptr = InStr(Ptr, Source, ",")
                            ' ***** 20140131
                            If return_ptr <= 0 Then
                                '
                                ' end of line, end of file
                                '
                                result = True
                            Else
                                return_ptr += 1
                                result = False
                            End If
                        End If
                        '
                        ' This is quoted text, so anything inside is converted to double quotes -- convert them back.
                        '
                        return_cell = Replace(return_cell, """""", """")
                        '
                    End If
                End If
                '
                ' determine eof
                '
                hint &= ",300"
                If return_ptr >= Len(Source) Then
                    return_eof = True
                End If
                Return result
            Catch ex As Exception
                Throw
            End Try
        End Function
        '
        '
        '
        Public Shared Sub parseLine(Source As String, source_ptr As Integer, ByRef return_cells() As String, ByRef return_ptr As Integer, ByRef return_eof As Boolean)
            Try
                '
                Dim Cell As String = ""
                Dim EOL As Boolean
                Dim fieldPtr As Integer
                '
                fieldPtr = 0
                return_ptr = source_ptr
                Do While Not EOL
                    Dim last As Integer
                    last = return_ptr
                    EOL = parseFieldReturnEol(Source, return_ptr, Cell, return_ptr, return_eof)
                    ReDim Preserve return_cells(fieldPtr)
                    return_cells(fieldPtr) = Cell
                    fieldPtr += 1
                    If return_ptr = 0 Then
                        return_ptr = return_ptr
                    End If
                Loop
            Catch ex As Exception
                Throw
            End Try
        End Sub
        '
        '
        '
        Public Shared Function parseFile(Source As String) As String(,)
            Try
                Dim result As String(,)
                '
                Dim EOL As Boolean
                Dim srcPtr As Integer
                Dim rowPtr As Integer
                Dim colPtr As Integer
                Dim eof As Boolean
                Dim colCnt As Integer
                Dim rowCnt As Integer
                Dim dummyCells As String() = Array.Empty(Of String)
                '
                ' parse the first row to get colCnt
                '
                Call parseLine(Source, 1, dummyCells, srcPtr, eof)
                colCnt = UBound(dummyCells) + 1
                rowCnt = 0
                '
                colPtr = 0
                rowPtr = 0
                srcPtr = 1
                ReDim result(colCnt - 1, 0)

                Do While Not eof
                    If rowPtr = 105 Then
                        rowPtr = rowPtr
                        If colPtr = 14 Then
                            colPtr = colPtr
                        End If
                    End If
                    If rowPtr >= rowCnt Then
                        rowCnt = rowPtr + 1
                        ReDim Preserve result(colCnt - 1, rowCnt - 1)
                    End If
                    If colPtr >= colCnt Then
                        If rowCnt <> 1 Then
                            '
                            ' error - can not adjust columns after first row
                            '
                        Else
                            colCnt = colPtr + 1
                            ReDim Preserve result(colCnt - 1, rowCnt - 1)
                        End If
                    End If
                    If (rowPtr = 79) And (colPtr = 2) Then
                        rowPtr = rowPtr
                    End If
                    Dim cell As String = ""
                    EOL = parseFieldReturnEol(Source, srcPtr, cell, srcPtr, eof)
                    result(colPtr, rowPtr) = cell

                    If EOL Then
                        colPtr = 0
                        rowPtr += 1
                    Else
                        If colPtr + 1 < colCnt Then
                            colPtr += 1
                        Else
                            colPtr = 0
                            rowPtr += 1
                        End If
                    End If
                Loop
                parseFile = result
                Return result
            Catch ex As Exception
                Throw
            End Try
        End Function
        '
        '
        '
        Public Shared Function firstNonZero(a As Integer, b As Integer) As Integer
            Dim v As Integer
            v = kmaGetFirstNonZeroLong(a, b)
            If v = 1 Then
                firstNonZero = a
            ElseIf v = 2 Then
                firstNonZero = b
            Else
                firstNonZero = 0
            End If
        End Function
        '
        Public Shared Function kmaGetFirstNonZeroLong(a As Integer, b As Integer) As Integer
            If (a = 0) And (b = 0) Then
                Return 0
            ElseIf (a = 0) Then
                Return 2
            ElseIf (b = 0) Or (a < b) Then
                Return 1
            Else
                Return 2
            End If
        End Function
    End Class
End Namespace

File: source/aoImportWizard/Controllers/GenericController.cs
using System;

namespace Contensive.ImportWizard.Controllers {
    public sealed class GenericController {
        private GenericController() {
        }
        //
        public static string normalizeFilename(string srcFilename) {
            string ext = System.IO.Path.GetExtension(srcFilename);
            string filenameNoExt = System.IO.Path.GetFileNameWithoutExtension(srcFilename);
            string result = "";
            string validCharacters = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
            foreach (var c in filenameNoExt.Trim().Replace(" ", "_")) {
                result += validCharacters.Contains(c.ToString()) ? c.ToString() : "-";
            }
            return string.IsNullOrEmpty(ext) ? result : $"{result}.{ext}";
        }
        //
        public static string Base64Encode(string plainText) {
            byte[] plainTextBytes = System.Text.Encoding.UTF8.GetBytes(plainText);
            return Convert.ToBase64String(plainTextBytes);
        }
        //
        public static string Base64Decode(string base64EncodedData) {
            byte[] base64EncodedBytes = Convert.FromBase64String(base64EncodedData);
            return System.Text.Encoding.UTF8.GetString(base64EncodedBytes);
        }
        //
        /// <summary>
        /// if date is invalid, set to minValue
        /// </summary>
        public static DateTime encodeMinDate(DateTime srcDate) {
            var returnDate = srcDate;
            if (srcDate < new DateTime(1900, 1, 1)) {
                returnDate = DateTime.MinValue;
            }
            return returnDate;
        }
        //
        /// <summary>
        /// if valid date, return the short date, else return blank string
        /// </summary>
        public static string getShortDateString(DateTime srcDate) {
            string returnString = "";
            var workingDate = encodeMinDate(srcDate);
            if (!isDateEmpty(srcDate)) {
                returnString = workingDate.ToShortDateString();
            }
            return returnString;
        }
        //
        public static bool isDateEmpty(DateTime srcDate) {
            return srcDate < new DateTime(1900, 1, 1);
        }
        //
        public static string getSortOrderFromInteger(int id) {
            return id.ToString().PadLeft(7, '0');
        }
        //
        public static string getDateForHtmlInput(DateTime source) {
            if (isDateEmpty(source)) {
                return "";
            } else {
                return $"{source.Year}-{source.Month.ToString().PadLeft(2, '0')}-{source.Day.ToString().PadLeft(2, '0')}";
            }
        }
        //
        public static string convertToDosPath(string sourcePath) {
            return sourcePath.Replace("/", @"\");
        }
        //
        public static string convertToUnixPath(string sourcePath) {
            return sourcePath.Replace(@"\", "/");
        }
        //
        // returns true if after removing this field, it is end of line
        // Returns a cell from a csv source and advances the ptr to the start of the next field
        // on entry, ptr points to the first character of the cell (1-based)
        // if at end of line, the parseFieldReturnEol is true
        // if end of file, return_eof is true
        //
        public static bool parseFieldReturnEol(string Source, int sourcePtr, ref string return_cell, ref int return_ptr, ref bool return_eof) {
            try {
                bool result;
                int crPtr;
                int endPtr;
                int Ptr;
                int workingPtr;
                bool IsQuoted;
                int commaptr;
                int lfPtr;
                int crlfPtr;
                //
                Ptr = sourcePtr;
                IsQuoted = false;
                result = false;
                //
                // find initial character
                //
                workingPtr = Ptr;
                while (Mid(Source, workingPtr, 1) == " ")
                    workingPtr += 1;
                if (Mid(Source, workingPtr, 1) == "\"") {
                    Ptr = workingPtr;
                    IsQuoted = true;
                }
                if (!IsQuoted) {
                    //
                    // non-Quoted field
                    //
                    commaptr = InStr(Ptr, Source, ",");
                    lfPtr = InStr(Ptr, Source, "\n");
                    crPtr = InStr(Ptr, Source, "\r");
                    crlfPtr = InStr(Ptr, Source, "\r\n");
                    //
                    // set workingPtr to the first one found
                    //
                    workingPtr = firstNonZero(commaptr, crlfPtr);
                    workingPtr = firstNonZero(workingPtr, lfPtr);
                    workingPtr = firstNonZero(workingPtr, crPtr);
                    workingPtr = firstNonZero(workingPtr, Source.Length);
                    if (workingPtr == crlfPtr) {
                        endPtr = workingPtr - 1;
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1);
                        return_ptr = workingPtr + 2;
                        result = true;
                    } else if (workingPtr == lfPtr) {
                        endPtr = workingPtr - 1;
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1);
                        return_ptr = workingPtr + 1;
                        result = true;
                    } else if (workingPtr == crPtr) {
                        endPtr = workingPtr - 1;
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1);
                        return_ptr = workingPtr + 1;
                        result = true;
                    } else if (workingPtr == commaptr) {
                        endPtr = workingPtr - 1;
                        return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1);
                        return_ptr = workingPtr + 1;
                        result = false;
                    } else {
                        endPtr = Source.Length;
                        if (endPtr - sourcePtr + 1 > 0) {
                            return_cell = Mid(Source, sourcePtr, endPtr - sourcePtr + 1);
                        } else {
                            return_cell = "";
                        }
                        return_ptr = endPtr;
                        result = true;
                    }
                } else {
                    //
                    // Quoted field, pass the initial quote
                    //
                    Ptr += 1;
                    int startPtr;
                    startPtr = Ptr;
                    //
                    while (Ptr != 0 & InStr(Ptr, Source, "\"") == InStr(Ptr, Source, "\"\"")) {
                        Ptr = InStr(Ptr, Source, "\"\"");
                        if (Ptr == 0) {
                            endPtr = Source.Length;
                            return_cell = "";
                            return_ptr = endPtr;
                            result = true;
                        } else {
                            Ptr += 2;
                        }
                    }
                    if (Ptr != 0) {
                        Ptr = InStr(Ptr, Source, "\"");
                        endPtr = Ptr - 1;
                        // skip white space to next delimiter
                        while (Mid(Source, Ptr + 1, 1) == " " & Ptr < Source.Length)
                            Ptr += 1;
                        if (Ptr >= Source.Length) {
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1);
                            return_ptr = Ptr + 3;
                            result = true;
                        } else if ((Mid(Source, Ptr + 1, 2) ?? "") == "\r\n") {
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1);
                            return_ptr = Ptr + 3;
                            result = true;
                        } else if ((Mid(Source, Ptr + 1, 1) ?? "") == "\n") {
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1);
                            return_ptr = Ptr + 2;
                            result = true;
                        } else if ((Mid(Source, Ptr + 1, 1) ?? "") == "\r") {
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1);
                            return_ptr = Ptr + 2;
                            result = true;
                        } else {
                            return_cell = Mid(Source, startPtr, endPtr - startPtr + 1);
                            return_ptr = InStr(Ptr, Source, ",");
                            if (return_ptr <= 0) {
                                result = true;
                            } else {
                                return_ptr += 1;
                                result = false;
                            }
                        }
                        //
                        // convert double quotes back to single quotes
                        //
                        return_cell = return_cell.Replace("\"\"", "\"");
                    }
                }
                //
                // determine eof
                //
                if (return_ptr >= Source.Length) {
                    return_eof = true;
                }
                return result;
            } catch (Exception) {
                throw;
            }
        }
        //
        public static void parseLine(string Source, int source_ptr, ref string[] return_cells, ref int return_ptr, ref bool return_eof) {
            try {
                string Cell = "";
                var EOL = default(bool);
                int fieldPtr;
                fieldPtr = 0;
                return_ptr = source_ptr;
                while (!EOL) {
                    int last;
                    last = return_ptr;
                    EOL = parseFieldReturnEol(Source, return_ptr, ref Cell, ref return_ptr, ref return_eof);
                    Array.Resize(ref return_cells, fieldPtr + 1);
                    return_cells[fieldPtr] = Cell;
                    fieldPtr += 1;
                    if (return_ptr == 0) {
                        break;
                    }
                }
            } catch (Exception) {
                throw;
            }
        }
        //
        public static string[,] parseFile(string Source) {
            try {
                string[,] result;
                bool EOL;
                var srcPtr = default(int);
                int rowPtr;
                int colPtr;
                var eof = default(bool);
                int colCnt;
                int rowCnt;
                string[] dummyCells = Array.Empty<string>();
                //
                // parse the first row to get colCnt
                //
                parseLine(Source, 1, ref dummyCells, ref srcPtr, ref eof);
                colCnt = dummyCells.GetUpperBound(0) + 1;
                rowCnt = 0;
                //
                colPtr = 0;
                rowPtr = 0;
                srcPtr = 1;
                result = new string[colCnt, 1];

                while (!eof) {
                    if (rowPtr >= rowCnt) {
                        rowCnt = rowPtr + 1;
                        var oldResult = result;
                        result = new string[colCnt, rowCnt];
                        if (oldResult is not null)
                            for (var i = 0; i <= oldResult.Length / oldResult.GetLength(1) - 1; ++i)
                                Array.Copy(oldResult, i * oldResult.GetLength(1), result, i * result.GetLength(1), Math.Min(oldResult.GetLength(1), result.GetLength(1)));
                    }
                    if (colPtr >= colCnt) {
                        if (rowCnt != 1) {
                            // error - can not adjust columns after first row
                        } else {
                            colCnt = colPtr + 1;
                            var oldResult1 = result;
                            result = new string[colCnt, rowCnt];
                            if (oldResult1 is not null)
                                for (var i1 = 0; i1 <= oldResult1.Length / oldResult1.GetLength(1) - 1; ++i1)
                                    Array.Copy(oldResult1, i1 * oldResult1.GetLength(1), result, i1 * result.GetLength(1), Math.Min(oldResult1.GetLength(1), result.GetLength(1)));
                        }
                    }
                    string cell = "";
                    EOL = parseFieldReturnEol(Source, srcPtr, ref cell, ref srcPtr, ref eof);
                    result[colPtr, rowPtr] = cell;

                    if (EOL) {
                        colPtr = 0;
                        rowPtr += 1;
                    } else if (colPtr + 1 < colCnt) {
                        colPtr += 1;
                    } else {
                        colPtr = 0;
                        rowPtr += 1;
                    }
                }
                return result;
            } catch (Exception) {
                throw;
            }
        }
        //
        public static int firstNonZero(int a, int b) {
            int v = kmaGetFirstNonZeroLong(a, b);
            if (v == 1) {
                return a;
            } else if (v == 2) {
                return b;
            } else {
                return 0;
            }
        }
        //
        public static int kmaGetFirstNonZeroLong(int a, int b) {
            if (a == 0 & b == 0) {
                return 0;
            } else if (a == 0) {
                return 2;
            } else if (b == 0 | a < b) {
                return 1;
            } else {
                return 2;
            }
        }
        //
        // Helper methods replacing VB Strings.Mid (1-based) and Strings.InStr (1-based)
        //
        /// <summary>
        /// VB-compatible Mid function (1-based index)
        /// </summary>
        private static string Mid(string source, int start, int length) {
            if (string.IsNullOrEmpty(source)) return "";
            int zeroStart = start - 1;
            if (zeroStart < 0) zeroStart = 0;
            if (zeroStart >= source.Length) return "";
            if (zeroStart + length > source.Length) length = source.Length - zeroStart;
            return source.Substring(zeroStart, length);
        }
        //
        /// <summary>
        /// VB-compatible InStr function (1-based index). Returns 0 if not found.
        /// </summary>
        private static int InStr(int start, string source, string find) {
            if (string.IsNullOrEmpty(source) || string.IsNullOrEmpty(find)) return 0;
            int zeroStart = start - 1;
            if (zeroStart < 0) zeroStart = 0;
            if (zeroStart >= source.Length) return 0;
            int idx = source.IndexOf(find, zeroStart, StringComparison.Ordinal);
            return idx < 0 ? 0 : idx + 1;
        }
    }
}
